Add Conveyer Subsystem

tails618 opened the pull request:
* Methods for the four actions we want the conveyer to do (hold the note, trap/amp score, shooter score, stop)
* Track the position of the note (currently using motor positions)
* Initialize beam break sensors as limit switches, in case the above doesn't work.

tails618 committed:
Some conveyer work

tails618 committed:
feat: Add relative encoder and limit switch functionality to Conveyer subsystem

This commit adds the following significant changes to the Conveyer subsystem:
- Import `RelativeEncoder` and `SparkLimitSwitch` classes
- Initialize relative encoders for front and back motors
- Initialize limit switch objects for beam break sensors
- Set up enum `NotePosition` with three possible values: NO_NOTE, RECEIVING_NOTE, HELD_NOTE
- Add method `startReceivingNote()` to set note position to RECEIVING_NOTE and update secondMotorZeroPosition
- Add method `getNotePosition()` to determine current note position based on motor movement and thresholds

The purpose of these changes is to enhance the functionality of the Conveyer subsystem by incorporating relative encoder readings and limit switch inputs. This will allow for more accurate tracking of note positions during operation.

avakimc commented on src/main/java/frc/robot/subsystems/conveyer/Conveyer.java:
```diff
@@ -0,0 +1,130 @@
+  public enum NotePosition {
+    NO_NOTE,
+    RECEIVING_NOTE,
+    HELD_NOTE
```
i would change HELD_NOTE to HOLDING_NOTE or something

avakimc reviewed (changes requested):
i think overall looks good

tails618 committed:
Rename HELD_NOTE to HOLDING_NOTE

avakimc reviewed (approved):
approved

balghane commented on src/main/java/frc/robot/subsystems/conveyer/Conveyer.java:
```diff
@@ -0,0 +1,130 @@
+    beamBreakSensorThree = backMotor.getForwardLimitSwitch(SparkLimitSwitch.Type.kNormallyOpen);
+    beamBreakSensorFour = backMotor.getReverseLimitSwitch(SparkLimitSwitch.Type.kNormallyOpen);
+
+    beamBreakSensorOne.enableLimitSwitch(false);
+    beamBreakSensorTwo.enableLimitSwitch(false);
+    beamBreakSensorThree.enableLimitSwitch(false);
+    beamBreakSensorFour.enableLimitSwitch(false);
```
These should be in the init function so you can check the returned revliberrors

balghane commented on src/main/java/frc/robot/subsystems/conveyer/Conveyer.java:
```diff
@@ -0,0 +1,130 @@
+  public Conveyer() {
+    SparkMaxUtils.initWithRetry(this::setUpConveyerSparks, ConveyerCal.SPARK_INIT_RETRY_ATTEMPTS);
+
+    beamBreakSensorOne = frontMotor.getForwardLimitSwitch(SparkLimitSwitch.Type.kNormallyOpen);
+    beamBreakSensorTwo = frontMotor.getReverseLimitSwitch(SparkLimitSwitch.Type.kNormallyOpen);
+    beamBreakSensorThree = backMotor.getForwardLimitSwitch(SparkLimitSwitch.Type.kNormallyOpen);
+    beamBreakSensorFour = backMotor.getReverseLimitSwitch(SparkLimitSwitch.Type.kNormallyOpen);
```
Can these just be set in the declarations above

balghane commented on src/main/java/frc/robot/subsystems/conveyer/Conveyer.java:
```diff
@@ -0,0 +1,130 @@
+    backMotor.set(ConveyerCal.SCORE_AMP_TRAP_BACK_SPEED);
+  }
+
+  public void hold() {
+    frontMotor.set(ConveyerCal.FRONT_HOLD_SPEED);
+    backMotor.set(ConveyerCal.BACK_HOLD_SPEED);
+  }
```
Receive?

balghane commented on src/main/java/frc/robot/subsystems/conveyer/Conveyer.java:
```diff
@@ -0,0 +1,130 @@
+    /**
+     * If the front motor is not moving and we currently have no note, we are not receiving a note
+     */
+    if (notePosition == NotePosition.NO_NOTE && frontMotorEncoder.getVelocity() < 1) {
```
split out 1. what units is that

balghane commented on src/main/java/frc/robot/subsystems/conveyer/Conveyer.java:
```diff
@@ -0,0 +1,130 @@
+   * Get the current position of the note. While we have the beam break sensors available, we are
+   * attempting to not use them, and rather just check the note position with the motors.
+   */
+  public NotePosition getNotePosition() {
```
When does this get called should it be in periodic or would it be called from a command

balghane replied:
The logic is a little confusing depending on when this gets called

tails618 replied:
This would be called in a command; I was thinking it would be like `WaitUntil(getNotePosition() == NotePosition.HOLDING_NOTE)` and then it would do the next thing

balghane commented on src/main/java/frc/robot/subsystems/conveyer/Conveyer.java:
```diff
@@ -0,0 +1,130 @@
+      notePosition = NotePosition.NO_NOTE;
+    } else {
+      /** If the back motor has moved more than the threshold, we have a note */
+      if (backMotorEncoder.getPosition() - secondMotorZeroPosition
+          > ConveyerCal.NOTE_THRESHOLD_DEGREES) {
```
Does get position return degrees?

tails618 committed:
Some work - temp push

tails618 committed:
More work on conveyer

tails618 committed:
Refactor conveyer subsystem and add new commands

- Refactored the `Conveyer` class to include new command imports.
- Added three new methods: `shoot()`, `scoreTrapOrAmp()`, and `receive()` which return instances of the respective commands.
- Updated the existing `stop()` method to return a command instance.
- Updated constants in `ConveyerCal` and `ConveyerConstants` classes with placeholder values.

Note: This commit refactors the conveyer subsystem by adding new commands for shooting, scoring in trap or amp, and receiving notes. It also updates some constants with placeholder values.

balghane commented on src/main/java/frc/robot/subsystems/conveyer/Conveyer.java:
```diff
@@ -0,0 +1,144 @@
+
+  public ConveyerPositions currentNotePosition = ConveyerPositions.NO_NOTE;
+
+  private double backMotorZeroPosition;
```
units

balghane commented on src/main/java/frc/robot/subsystems/conveyer/Conveyer.java:
```diff
@@ -0,0 +1,144 @@
+        SparkMaxUtils.check(
+            backMotor.setSmartCurrentLimit(ConveyerCal.CONVEYER_CURRENT_LIMIT_AMPS));
+
+    errors +=
+        SparkMaxUtils.check(
+            SparkMaxUtils.UnitConversions.setDegreesFromGearRatio(frontMotorEncoder, ConveyerConstants.FRONT_GEAR_RATIO));
+    errors +=
+        SparkMaxUtils.check(
+            SparkMaxUtils.UnitConversions.setDegreesFromGearRatio(backMotorEncoder, ConveyerConstants.BACK_GEAR_RATIO));
```
Should we set these to linear? I landed some stuff for setting these to linear for the elevator

balghane commented on src/main/java/frc/robot/subsystems/conveyer/Conveyer.java:
```diff
@@ -0,0 +1,144 @@
+    return errors == 0;
+  }
+
+  public Command shoot() {
```
Remember to add requirements

balghane commented on src/main/java/frc/robot/subsystems/conveyer/Conveyer.java:
```diff
@@ -0,0 +1,144 @@
+      new WaitCommand(ConveyerCal.NOTE_EXIT_TIME_SHOOTER_SECONDS),
+      stop(),
+      new InstantCommand(() -> currentNotePosition = ConveyerPositions.NO_NOTE),
+      new WaitUntilCommand(() -> backMotorEncoder.getVelocity() < ConveyerCal.MOTOR_VELOCITY_THRESHOLD_RPM),
+      new InstantCommand(() -> backMotorZeroPosition = backMotorEncoder.getPosition())
```
Hm why not set this at the start of receive? What if it shifts a little bit

balghane commented on src/main/java/frc/robot/subsystems/conveyer/Conveyer.java:
```diff
@@ -0,0 +1,144 @@
+      new InstantCommand(() -> currentNotePosition = ConveyerPositions.PARTIAL_NOTE),
+      new WaitUntilCommand(() -> Math.abs(backMotorEncoder.getPosition() - backMotorZeroPosition) < ConveyerCal.NOTE_POSITION_THRESHOLD_DEGREES),
+      new InstantCommand(() -> backMotorZeroPosition = backMotorEncoder.getPosition()),
+      new InstantCommand(() -> frontMotor.set(ConveyerCal.BACK_OFF_POWER)),
+      new InstantCommand(() -> backMotor.set(ConveyerCal.BACK_OFF_POWER)),
+      new WaitUntilCommand(() -> Math.abs(backMotorEncoder.getPosition() - backMotorZeroPosition) < ConveyerCal.BACK_OFF_DEGREES),
```
Maybe use front position for this? Back might slip on the note slightly but front def will not. Good idea tho

tails618 committed:
feat: Update Conveyer class and constants

- Updated the variable names `backMotorZeroPosition` to `backMotorZeroPositionMeters` and `frontMotorZeroPosition` to `frontMotorZeroPositionMeters` in the Conveyer class.
- Modified the method `shoot`, `scoreTrapOrAmp`, and `receive` in the Conveyer class to accept a parameter of type Conveyer.
- Updated the usage of variables in the methods `shoot`, `scoreTrapOrAmp`, and `receive` in the Conveyer class to use the passed conveyer object instead of accessing them directly.
- Renamed constant variable `NOTE_POSITION_THRESHOLD_DEGREES` to `NOTE_POSITION_THRESHOLD_IN`.
- Renamed constant variable `BACK_OFF_DEGREES` to `BACK_OFF_IN`.
- Added a new constant variable named CONVEYER_MOTOR_ROLLER_DIAMETER_INS.

balghane commented on src/main/java/frc/robot/subsystems/conveyer/ConveyerConstants.java:
```diff
@@ -0,0 +1,13 @@
+  public static final double FRONT_GEAR_RATIO = Constants.PLACEHOLDER_DOUBLE,
+    BACK_GEAR_RATIO = Constants.PLACEHOLDER_DOUBLE;
+  
+  public static final double CONVEYER_MOTOR_ROLLER_DIAMETER_INS = Constants.PLACEHOLDER_DOUBLE;
```
Just IN for inches. or INCHES

tails618 replied:
Changed

balghane commented on src/main/java/frc/robot/subsystems/conveyer/ConveyerCal.java:
```diff
@@ -0,0 +1,35 @@
+  public static final double NOTE_POSITION_THRESHOLD_IN = 0.5;
+
+  /** The threshold required for the velocity of a motor in order to count it as moving */
+  public static final double MOTOR_VELOCITY_THRESHOLD_RPM = 1.0;
```
Maybe 10 or something?

balghane replied:
But idk what it would be in inches

tails618 replied:
Placeholder for now. I also have to rename this var.

tails618 replied:
Changed

balghane commented on src/main/java/frc/robot/subsystems/conveyer/Conveyer.java:
```diff
@@ -0,0 +1,143 @@
+      new WaitCommand(ConveyerCal.NOTE_EXIT_TIME_SHOOTER_SECONDS),
+      Conveyer.stop(conveyer),
+      new InstantCommand(() -> conveyer.currentNotePosition = ConveyerPositions.NO_NOTE),
+      new WaitUntilCommand(() -> conveyer.backMotorEncoder.getVelocity() < ConveyerCal.MOTOR_VELOCITY_THRESHOLD_RPM)
```
Now that you changed the encoder to linear, this velocity will also be linear

tails618 replied:
Changed

tails618 commented on src/main/java/frc/robot/subsystems/conveyer/Conveyer.java:
```diff
@@ -0,0 +1,143 @@
+package frc.robot.subsystems.conveyer;
```
I forgot to add stuff to shuffleboard

tails618 replied:
Added

tails618 committed:
feat: Add Shuffleboard properties for conveyer subsystem

This commit adds new Shuffleboard properties to the Conveyer subsystem. The "Current Note Position" property displays the current position of the note, while the "Front Motor Zero Position (in)" and "Back Motor Zero Position (in)" properties show the zero positions of the front and back motors in inches. Additionally, properties for motor positions and velocities are included.

Changes:
- Added SendableBuilder import
- Implemented initSendable() method in Conveyer class
- Added new Shuffleboard properties
